compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp

```diff
@@ -0,0 +1,614 @@
+      {rs.originalFillH ? rs.originalFillH : originalFillH, rs.splitFillH,
+       rs.splitLinalgH});
+  if (!optionalFusionRootH) {
+    optionalFusionRootH = rs.combinerH;
+  } else {
+    optionalFusionRootH =
+        rs.trailingEltwiseH ? rs.trailingEltwiseH : optionalFusionRootH;
+    opsHToFuse.push_back(rs.combinerH);
+  }
+  if (rs.leadingEltwiseH) {
+    opsHToFuse.push_back(rs.leadingEltwiseH);
+  }
+
+  // The presence of leading elementwise operation implies that dispatch region
+  // formation happened using another transform dialect script and doesn't need
+  // the workgroup count part.
+  if (hasLeadingEltwise) {
+    buildTileFuseDistWithTileSizes<TileToForeachThreadOp>(
+        b, optionalFusionRootH, opsHToFuse, tileSizes0Generic,
+        b.getArrayAttr({x}));
+  } else {
+    buildTileFuseDistWithTileSizes<
+        TileToForeachThreadAndWorkgroupCountRegionOp>(
+        b, optionalFusionRootH, opsHToFuse, tileSizes0Generic,
+        b.getArrayAttr({x}));
+  }
+
+  return variantH;
+}
+
+static Value buildReductionStrategyThreadDistributionPart(
+    ImplicitLocOpBuilder &b, Value variantH, ArrayRef<int64_t> tileSizes1Fill,
+    ArrayRef<int64_t> tileSizes1Generic, bool hasLeadingEltwise,
+    bool hasTrailingEltwise) {
+  // TODO: Relying on ordering is brittle, harden this.
+  Value matchedH = b.create<MatchOp>(
+      variantH, ArrayRef<StringRef>{linalg::GenericOp::getOperationName(),
+                                    linalg::FillOp::getOperationName()});
+  auto split = b.create<SplitHandlesOp>(
+      matchedH,
+      /*numResultHandles=*/4 + hasLeadingEltwise + hasTrailingEltwise);
+  Value firstFusionRootH = split.getResults()[1 + hasLeadingEltwise];
+  SmallVector<Value> firstFusionGroupHs =
+      split.getResults().take_front(1 + hasLeadingEltwise);
+  Value secondFusionRootH = split.getResults().back();
+  SmallVector<Value> secondFusionGroupHs =
+      split.getResults().drop_front(2 + hasLeadingEltwise).drop_back();
+
+  auto z = mlir::gpu::GPUThreadMappingAttr::get(b.getContext(),
+                                                ::mlir::gpu::Threads::DimZ);
+  auto y = mlir::gpu::GPUThreadMappingAttr::get(b.getContext(),
+                                                ::mlir::gpu::Threads::DimY);
+
+  // clang-format off
+  buildTileFuseDistWithTileSizes<TileToForeachThreadOp>(b,
+                   /*rootH=*/secondFusionRootH,
+                   /*opsHToFuse=*/secondFusionGroupHs,
+                   /*tileSizes=*/tileSizes1Fill,
+                   /*threadDimMapping=*/b.getArrayAttr({z}));
+  buildTileFuseDistWithTileSizes<TileToForeachThreadOp>(b,
+                   /*rootH=*/firstFusionRootH,
+                   /*opsHToFuse=*/firstFusionGroupHs,
+                   /*tileSizes=*/tileSizes1Generic,
+                   /*threadDimMapping=*/b.getArrayAttr({z,y}));
+  // clang-format on
+  return variantH;
+}
+
+/// Structure to hold the parameters related to GPU reduction strategy.
```

Comment:
The code below this line should be factored out to LLVMCPU/ and LLVMGPU/. They are not common utils, but they live in Coomon/. Some of the codes are CPU specific and some of them are CUDA specific.

Reply:
This falls into the category of trivial things that take time and can be separated in a followup commit as was proposed by @ThomasRaoux .

Is this suggestion intended to be a blocker for landing or is a followup move-only PR the way forward?